From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Juhyung Park <qkrwngud825@gmail.com>
Date: Fri, 22 Mar 2024 11:38:02 +0900
Subject: [PATCH 29/34] bionic: use jemalloc by default and allow opt-in to
 scudo (3/4)

Overall, jemalloc performs 23.4% better than Scudo while consuming
1.7% less amount of memory in system/extra's real-world memory_replay
traces:

https://docs.google.com/spreadsheets/d/19wLanILSw_l2d7oot3VvcZiFSLzPRWNb_a5ZyA3aFSU/edit?usp=sharing

Scudo performs fairly well for a hardened memory allocator, but we're
optimizing for performance.

Tests were performed by statically linking memory_replay under QSSI 14.
Android userspace has been stopped before performing all benchmarks,
and each test was executed 10 times with an idle detector in between to
reduce noise and errors.

The entire test took about 3 days:
https://github.com/arter97/android_memory_replay_helper

This commit has been inspired from https://github.com/ProtonAOSP/android_bionic/commit/b2204890eac02f54738a8d9e17c93643ff0aff60

Change-Id: I5f6955bac87d87f9ef931c9be2fd43ac5d8fbcd7
Signed-off-by: Juhyung Park <qkrwngud825@gmail.com>
---
 libc/Android.bp | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/libc/Android.bp b/libc/Android.bp
index a590dac..bc04e03 100644
--- a/libc/Android.bp
+++ b/libc/Android.bp
@@ -99,6 +99,9 @@ cc_defaults {
         malloc_low_memory: {
             cflags: ["-UUSE_SCUDO"],
         },
+        malloc_use_jemalloc: {
+            cflags: ["-UUSE_SCUDO"],
+        },
     },
 
     lto: {
@@ -191,6 +194,16 @@ cc_defaults {
                 "libscudo",
             ],
         },
+        malloc_use_jemalloc: {
+            cflags: ["-UUSE_SCUDO"],
+            whole_static_libs: [
+                "libjemalloc5",
+                "libc_jemalloc_wrapper",
+            ],
+            exclude_static_libs: [
+                "libscudo",
+            ],
+        },
     },
 }
 
